library/std/src/sys/unix/locks/queue_rwlock.rs

```diff
@@ -0,0 +1,492 @@
+const LOCKED: usize = 1;
+const QUEUED: usize = 2;
+const SINGLE: usize = 4;
+const MASK: usize = !(LOCKED | QUEUED);
+
+/// Returns a closure that changes the state to the lock state corresponding to
+/// the lock mode indicated in `read`.
+#[inline]
+fn lock(read: bool) -> impl Fn(State) -> Option<State> {
+    move |state| {
+        if read {
+            if state.addr() & QUEUED == 0 && state.addr() != LOCKED {
+                Some(invalid_mut(state.addr().checked_add(SINGLE)? | LOCKED))
+            } else {
+                None
+            }
+        } else {
+            let state = state.wrapping_byte_add(LOCKED);
+            if state.addr() & LOCKED == LOCKED { Some(state) } else { None }
+        }
+    }
+}
+
+/// Masks the state, assuming it points to a queue node.
+///
+/// # Safety
+/// The state must contain a valid pointer to a queue node.
+#[inline]
+unsafe fn to_node(state: State) -> NonNull<Node> {
+    unsafe { NonNull::new_unchecked(state.mask(MASK)).cast() }
+}
+
+/// An atomic node pointer with relaxed operations.
+struct AtomicLink(AtomicPtr<Node>);
+
+impl AtomicLink {
+    fn new(v: Option<NonNull<Node>>) -> AtomicLink {
+        AtomicLink(AtomicPtr::new(v.map_or(null_mut(), NonNull::as_ptr)))
+    }
+
+    fn get(&self) -> Option<NonNull<Node>> {
+        NonNull::new(self.0.load(Relaxed))
+    }
+
+    fn set(&self, v: Option<NonNull<Node>>) {
+        self.0.store(v.map_or(null_mut(), NonNull::as_ptr), Relaxed);
+    }
+}
+
+#[repr(align(4))]
+struct Node {
+    next: AtomicLink,
+    prev: AtomicLink,
+    tail: AtomicLink,
+    read: bool,
+    thread: OnceCell<Thread>,
+    completed: AtomicBool,
+}
+
+impl Node {
+    /// Create a new queue node.
+    fn new(read: bool) -> Node {
+        Node {
+            next: AtomicLink::new(None),
+            prev: AtomicLink::new(None),
+            tail: AtomicLink::new(None),
+            read,
+            thread: OnceCell::new(),
+            completed: AtomicBool::new(false),
+        }
+    }
+
+    /// Set the `next` field depending on the lock state. If there are threads
+    /// queued, the `next` field will be set to a pointer to the next node in
+    /// the queue. Otherwise the `next` field will be set to the lock count if
+    /// the state is read-locked or to zero if it is write-locked.
+    fn set_state(&mut self, state: State) {
+        self.next.0 = AtomicPtr::new(state.mask(MASK).cast());
+    }
+
+    /// Assuming the node contains a reader lock count, decrement that count.
+    /// Returns `true` if there are other lock owners.
+    fn decrement_count(&self) -> bool {
+        self.next.0.fetch_byte_sub(SINGLE, AcqRel).addr() > SINGLE
+    }
+
+    /// Prepare this node for waiting.
+    fn prepare(&mut self) {
+        // Fall back to creating an unnamed `Thread` handle to allow locking in
+        // TLS destructors.
+        self.thread.get_or_init(|| thread_info::current_thread().unwrap_or(Thread::new(None)));
```

Comment:
Should be `unwrap_or_else` to save Thread::new call?

Reply:
Oh yes, that's right!